karapace/kafka_admin.py

```diff
@@ -0,0 +1,290 @@
+"""
+Copyright (c) 2023 Aiven Ltd
+See LICENSE for details
+"""
+
+from __future__ import annotations
+
+from collections.abc import Iterable
+from concurrent.futures import Future
+from confluent_kafka import TopicPartition
+from confluent_kafka.admin import (
+    AdminClient,
+    BrokerMetadata,
+    ClusterMetadata,
+    ConfigResource,
+    ConfigSource,
+    NewTopic,
+    OffsetSpec,
+    ResourceType,
+    TopicMetadata,
+)
+from confluent_kafka.error import KafkaException
+from kafka.errors import AuthenticationFailedError, for_code, UnknownTopicOrPartitionError
+from karapace.constants import TOPIC_CREATION_TIMEOUT_S
+from typing import Any, Callable, Container, NoReturn, Protocol, TypedDict, TypeVar
+from typing_extensions import Unpack
+
+import logging
+
+LOG = logging.getLogger(__name__)
+
+
+T = TypeVar("T")
+
+
+def single_futmap_result(futmap: dict[Any, Future[T]]) -> T:
+    """Extract the result of a future wrapped in a dict.
+
+    Bulk operations of the `confluent_kafka` library's Kafka clients return results
+    wrapped in a dictionary of futures. Most often we use these bulk operations to
+    operate on a single resource/entity. This function makes sure the dictionary of
+    futures contains a single future and returns its result.
+    """
+    assert len(futmap) == 1
```

Comment:
```suggestion
    operate on a single resource/entity. This function makes sure the dictionary of
    futures contains a single future and returns its result.
    """
    if not len(futmap) == 1:
        raise ValueError
```

suggestion: If it's part of the functions responsibility to "make sure the dictionary of futures contains a single future", I don't think it's appropriate with an assertion.

Reply:
(golfing) something would this could be a lazy way to make sure we raise when there's not exactly one item:

```python
(future,) = futmap.values()
return future.result()
```